### PR TITLE
(Civis) Updating Socrata Import

### PR DESCRIPTION
**Pull request by akoebs**
Updated the import_socrata.py and helpers.py files to reflect current versions in Platform.  A newer, improved version should be coming soon as well with better handling of data-types after Ian discovered the dates issue.  Once this is merged to master we will switch the Github reference in Platform to reference the Aqueduct repo on the core script and any changes will only be referenced there from that point forward.

### CODE REVIEW

File: civis/imports/socrata_helpers.py [hunk 1]
```diff
@@ -125,30 +179,51 @@ def _store_and_attach_metadata(
...
 def _read_paginated(
-    client,
-    dataset_id: int,
-    page_limit: int = 50000,
-    size_limit: int = None,
-    where: str = None,
+    client, dataset_id: int, page_limit: int = 90000, size_limit: int = None
 ):
+    """
+    Uses socrata API get call to pull i.
```
Comment (ian-r-rose) [inline civis/imports/socrata_helpers.py:185]: Typo?

File: civis/imports/socrata_helpers.py [hunk 2]
```diff
@@ -125,30 +179,51 @@ def _store_and_attach_metadata(
...
+    Pulls in Socrata data in a while loop using pandas dataframes (in 80000 row
+    chunks) that get written to .csv(s). .csv(s) and then appened together and
+    a path to final csv is outputted.
+    Inputs
+    _____________
+    dataset_id: str
+        Socrata dataset identifier
+    page_limit: int
+        number of records that can be pulled in chunk
+    date_column: 'str'
```
Comment (ian-r-rose) [inline civis/imports/socrata_helpers.py:195]: These documented arguments (especially `date_column`) don't seem to match the actual arguments.

File: civis/imports/socrata_helpers.py [hunk 3]
```diff
@@ -167,4 +242,153 @@ def _read_paginated(
...
+    """
+    Uses socrata get calls to query data in evenly spaced chunks (by date)
+    writes to pd.df then to csv. All csvs then appended into one large csv.
+    only one chunk in memory at time, as csv is on disk.
+    Inputs
+    _____________
+    dataset_id: str
+        Socrata dataset identifier
+    page_limit: int
+        number of records that can be pulled in chunk
```
Comment (ian-r-rose) [inline civis/imports/socrata_helpers.py:281]: Size limit is missing here.

File: civis/imports/socrata_helpers.py [hunk 4]
```diff
@@ -167,4 +242,153 @@ def _read_paginated(
...
+    _____________
+    min_date: str
+        Oldest date in dataset
+    max_date: str
+        most recent date in dataset
+    date_splits: array
+        array of n dates set by splits param.
+    """
+    min_date_query = (
+        "SELECT " + date_column + " Order by " + date_column + " ASC limit 1"
```
Comment (ian-r-rose) [inline civis/imports/socrata_helpers.py:372]: Seems correct, but is there a reason why `max(date_column)` doesn't work here and below?

File: civis/imports/socrata_helpers.py [hunk 5]
```diff
@@ -21,35 +23,88 @@ def _parse_metadata(metadata: dict, paths: dict):
         out[name] = reduce(
             lambda d, key: d.get(key, None) if isinstance(d, dict) else None,
             path.split("."),
-            metadata,
+            metadata
         )
     return out
 
 
+def _store_and_attach_metadata(
```
Comment (ian-r-rose) [inline civis/imports/socrata_helpers.py:31]: What is the purpose of this function? What are the raw and cleaned metadata, and who is looking at them?

File: civis/imports/socrata_helpers.py (hunk 4 above)
Comment (snassef) [reply civis/imports/socrata_helpers.py:372]: SoQl ( the socrata query engine) doesn't support aggregate functions. We ended up moving away from this method to a more efficient one using page_limits and offset which is not data set dependent. I kept this in for reference but can delete.

File: civis/imports/socrata_helpers.py (hunk 5 above)
Comment (snassef) [reply civis/imports/socrata_helpers.py:31]: Raw metadata contains description, author, columns, column type, min max vals etc. We are using raw metadata (in the new import) to map Socrata datatypes to PostGIS/Redshift datatypes. 

Clean metadata pretty much pulls the description out of metadata and attaches it to the table in platfrom as a description.

File: civis/imports/socrata_helpers.py (hunk 3 above)
Comment (snassef) [reply civis/imports/socrata_helpers.py:281]: Thanks for catching these -

File: civis/imports/socrata_helpers.py (hunk 5 above)
Comment (ian-r-rose) [reply civis/imports/socrata_helpers.py:31]: > 
> Raw metadata contains description, author, columns, column type, min max vals etc. We are using raw metadata (in the new import) to map Socrata datatypes to PostGIS/Redshift datatypes.

Interesting -- where is the code doing this? How does uploading a metadata file to s3 help accomplish getting the datatypes right in the dbs? This is all still pretty opaque to me, my instinct would be to create the create the table with the appropriate data types in this same script, using standard SQL or an ORM.

> 
> Clean metadata pulls the description out of raw metadata and attaches it to the table in platfrom as a description.

:+1:

File: civis/imports/socrata_helpers.py (hunk 4 above)
Comment (ian-r-rose) [reply civis/imports/socrata_helpers.py:372]: If it's not used, I'd just as soon delete it. We can always go into git history if we need it.

**Comment by hunterowens** [comment]
@snassef any chance to get a follow on commit here removing some of the dead code?

**Comment by hunterowens** [comment]
@snassef @akoebs following up on the extra commits here

**Comment by akoebs** [comment]
@hunterowens @ian-r-rose extra/new commits added!

File: civis/imports/socrata_helpers.py (hunk 5 above)
Comment (snassef) [reply civis/imports/socrata_helpers.py]: The new code that was just submitted has better documentation and should clear things up. In the meeting today I will go over, at a high level, the code.

Comment (snassef) [reply civis/imports/socrata_helpers.py]: image
_store_and_attach_metadata allows us to attach metadata as a JSON (which is stored in S3) to the output of the run and also attaches the Socrata description and author to the table that the data is being stored in. The datatype mapping is not occurring in  _store_and_attach_metadata. It occurs in create_col_type_dict which you can find in the latest commits.

**Pull request merged**